### ISSUE TITLE
The IL compiler failed to compile due to an unexpected error. Stack empty.

### ISSUE THREAD

**Issue by DaveOsDev** [FlingOS/FlingOS#86]
Error       The IL compiler failed to compile due to an unexpected error. Stack empty.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   at System.Collections.Generic.Stack`1.Pop()
   at Drivers.Compiler.Architectures.x86.Callvirt.PerformStackOperations(ILPreprocessState conversionState, ILOp theOp) in C:\Users\DaveOSDev\Source\Repos\FlingOS\Drivers\Compiler\Architectures\x86_32\ILOps\Callvirt.cs:line 101
   at Drivers.Compiler.IL.ILPreprocessor.InjectGC(ILLibrary TheLibrary, MethodInfo theMethodInfo, ILBlock theILBlock) in C:\Users\DaveOSDev\Source\Repos\FlingOS\Drivers\Compiler\IL\ILPreprocessor.cs:line 909
   at Drivers.Compiler.IL.ILPreprocessor.Preprocess(ILLibrary TheLibrary) in C:\Users\DaveOSDev\Source\Repos\FlingOS\Drivers\Compiler\IL\ILPreprocessor.cs:line 85
   at Drivers.Compiler.IL.ILPreprocessor.Preprocess(ILLibrary TheLibrary) in C:\Users\DaveOSDev\Source\Repos\FlingOS\Drivers\Compiler\IL\ILPreprocessor.cs:line 71
   at Drivers.Compiler.IL.ILCompiler.ExecuteILPreprocessor(ILLibrary TheLibrary) in C:\Users\DaveOSDev\Source\Repos\FlingOS\Drivers\Compiler\IL\ILCompiler.cs:line 113
   at Drivers.Compiler.IL.ILCompiler.Compile(ILLibrary TheLibrary) in C:\Users\DaveOSDev\Source\Repos\FlingOS\Drivers\Compiler\IL\ILCompiler.cs:line 62
   at Drivers.Compiler.App.CompilerProcess.Execute(LogMessageEventHandler messageHandler, LogWarningEventHandler warningHandler, LogErrorEventHandler errorHandler) in C:\Users\DaveOSDev\Source\Repos\FlingOS\Drivers\Compiler\App\CompilerProcess.cs:line 238 Kernel (Kernel\Kernel)  C:\Users\DaveOSDev\Source\Repos\FlingOS\Kernel\Kernel\Kernel.csproj 175

**Comment by EdNutting** [comment]
Okay so this has been introduced by a change in 957b8a4 . The IL code for the chunk I'm referring to looks like this:

```
.method public hidebysig static 
    void Clean () cil managed 
{
    // Method begins at RVA 0x11fec
    // Code size 19 (0x13)
    .maxstack 8

    IL_0000: nop
    IL_0001: ldsfld class Kernel.Hardware.Keyboards.PS2 Kernel.Hardware.Keyboards.PS2::ThePS2
    IL_0006: dup
    IL_0007: brtrue.s IL_000c

    IL_0009: pop
    IL_000a: br.s IL_0012

    IL_000c: callvirt instance void Kernel.Hardware.Devices.Keyboard::Disable()
    IL_0011: nop

    IL_0012: ret
} // end of method PS2::Clean
```

(Obtained using ILSpy - ILSpy is a great friend in this kind of development work).

Unfortunately, this is due to an oversight in my compiler design. I forgot that branch operations cause the stack to fork. What this means is the virtual stack the compiler uses to track stuff is fine right up until the first branch op. After that point, it does the Pop operation, which empties the stack and then continues straight on to the CallVirt op, which then causes the stack empty error.

I will work on a fix for this today / over the next few days. For now, you can simply not use the "?" operator in order to avoid the bug.

**Comment by EdNutting** [comment]
Resolved in 40e8647. I'll merge `EdNutting` into `develop` then you can merge down from `develop` later.
